lib/elixir/lib/map.ex

```diff
@@ -93,11 +96,10 @@ defmodule Map do
   """
   @spec new(Enum.t, (term -> {key, value})) :: map
   def new(enumerable, transform) do
-    fun = fn el, acc ->
-      {k, v} = transform.(el)
-      put(acc, k, v)
-    end
-    Enum.reduce(enumerable, %{}, fun)
+    enumerable
+    |> Enum.reduce([], &[transform.(&1) | &2])
+    |> :lists.reverse
+    |> :maps.from_list
```

Comment:
From what I've seen we don't do any optimizations for Enum.into an empty (or very small) map. I do believe that the best place for optimization is the `Enum.into` module itself, rather than the collectable protocol implementation for maps or mapsets.